lucene/grouping/src/java/org/apache/lucene/search/grouping/FirstPassGroupingCollector.java

```diff
@@ -132,17 +132,28 @@ public ScoreMode scoreMode() {
     final Collection<SearchGroup<T>> result = new ArrayList<>();
     int upto = 0;
     final int sortFieldCount = comparators.length;
+    assert sortFieldCount > 0; // this must always be true because fields Sort must contain at least a field
     for(CollectedSearchGroup<T> group : orderedGroups) {
       if (upto++ < groupOffset) {
         continue;
       }
       // System.out.println("  group=" + (group.groupValue == null ? "null" : group.groupValue.toString()));
       SearchGroup<T> searchGroup = new SearchGroup<>();
       searchGroup.groupValue = group.groupValue;
+      // We pass this around so that we can get the corresponding solr id when serializing the search group to send to the federator
+      searchGroup.topDocLuceneId = group.topDoc;
       searchGroup.sortValues = new Object[sortFieldCount];
       for(int sortFieldIDX=0;sortFieldIDX<sortFieldCount;sortFieldIDX++) {
         searchGroup.sortValues[sortFieldIDX] = comparators[sortFieldIDX].value(group.comparatorSlot);
       }
+      searchGroup.topDocScore = Float.NaN;
+      // if there is the score comparator we want to return the score
+      for (FieldComparator comparator: comparators){
+        if (comparator instanceof FieldComparator.RelevanceComparator){
+          searchGroup.topDocScore = (Float)comparator.value(group.comparatorSlot);
+        }
```

Comment:
The `searchGroup.topDocScore` being filled in conditionally here i.e. only if it's a field comparator 'jumps out' a little.

Have not yet looked into what other comparators there are and if removing of the conditional might be practical.

But this also re-surfaced the `SearchGroup.java` question around adding or not adding additional fields and what (if any) alternative might be possible.

With this in mind https://github.com/cpoerschke/lucene-solr/commits/github-bloomberg-SOLR-11831-cpoerschke-13 explores the 'extend FirstPassGroupingCollector and SearchGroup' route a little further:
 * presume LUCENE-8728 changes or equivalent are available
 * add SolrFirstPassGroupingCollector and SolrSearchGroup classes
 * on the `group.skip.second.step` code paths always have SolrSearchGroup instead of SearchGroup
 * not yet done:
   * creation of SolrFirstPassGroupingCollector instead of FirstPassGroupingCollector if-and-only-if appropriate
   * tests need to pass again
   * consider if this would actually be comprehensible, maintainable and helpful e.g. w.r.t. (Solr)SearchGroup fields always being filled in.

What do you think?

Reply:
Taking the conversation back from the commits on the above branch to here: https://github.com/bloomberg/lucene-solr/pull/232 sketches extending Lucene's FirstPassGroupingCollector and SearchGroup for Solr's group.skip.second.step use as outlined above.

Flagged up by that sketching, and not yet fully explored, is the question of whether or not the exclusion of group.func is actually needed, superficially it seemed it might not be i.e. the federator is just merging groups and sorting by id(s) and where the ids came from might not matter.